libnetwork/network_windows.go

```diff
@@ -66,6 +77,144 @@ func (n *Network) startResolver() {
+		return nil
+	}
+
+	// Get the list of DNS servers HNS has set up for this Endpoint.
+	var dnsList []extDNSEntry
+	dnsServers := strings.Split(hnsEp.DNSServerList, ",")
+
+	// Create an extDNSEntry for each DNS server, apart from 'resolver' itself.
+	var foundSelf bool
+	for _, dnsServer := range dnsServers {
+		if sameAddrs(dnsServer, hnsEp.GatewayAddress) || sameAddrs(dnsServer, hnsEp.GatewayAddressV6) {
+			foundSelf = true
+		} else {
+			dnsList = append(dnsList, extDNSEntry{IPStr: dnsServer})
+		}
+	}
+	if !foundSelf {
+		log.G(context.TODO()).WithFields(log.Fields{
+			"network":  netName,
+			"endpoint": epName,
+		}).Debug("Endpoint is not configured to use internal DNS resolver")
+		return nil
+	}
+
+	// If the internal resolver is configured as one of this endpoint's DNS servers,
+	// tell it which ext servers to use for requests from this endpoint's addresses.
+	s := make([]string, len(dnsList))
+	for i, d := range dnsList {
+		s[i] = d.IPStr
+	}
+	log.G(context.TODO()).Infof("External DNS servers for '%s': %v", epName, s)
+	if srcAddr, ok := netip.AddrFromSlice(hnsEp.IPAddress); ok {
+		if err := resolver.SetExtServersForSrc(srcAddr.Unmap(), dnsList); err != nil {
+			return errors.Wrapf(err, "failed to set external DNS servers for %s address %s",
+				epName, hnsEp.IPAddress)
+		}
+	}
+	if srcAddr, ok := netip.AddrFromSlice(hnsEp.IPv6Address); ok {
+		if err := resolver.SetExtServersForSrc(srcAddr, dnsList); err != nil {
+			return errors.Wrapf(err, "failed to set external DNS servers for %s address %s",
+				epName, hnsEp.IPv6Address)
+		}
+	}
+	return nil
+}
+
+func deleteEpFromResolver(epName string, epIface *EndpointInterface, resolvers []*Resolver) error {
+	// Find the HNSEndpoint represented by ep, matching on endpoint address.
+	hnsEp := findHNSEp(epIface.addr, epIface.addrv6)
+	if hnsEp == nil {
+		return fmt.Errorf("no HNS endpoint for %s", epName)
+	}
+
+	// Find the resolver for that HNSEndpoint, matching on gateway address.
+	resolver := findResolver(resolvers, hnsEp.GatewayAddress, hnsEp.GatewayAddressV6)
+	if resolver == nil {
+		return nil
+	}
+
+	// Delete external DNS servers for the endpoint's IP addresses.
+	if srcAddr, ok := netip.AddrFromSlice(hnsEp.IPAddress); ok {
+		if err := resolver.SetExtServersForSrc(srcAddr.Unmap(), nil); err != nil {
+			return errors.Wrapf(err, "failed to delete external DNS servers for %s address %s",
+				epName, hnsEp.IPv6Address)
+		}
+	}
+	if srcAddr, ok := netip.AddrFromSlice(hnsEp.IPv6Address); ok {
+		if err := resolver.SetExtServersForSrc(srcAddr, nil); err != nil {
+			return errors.Wrapf(err, "failed to delete external DNS servers for %s address %s",
+				epName, hnsEp.IPv6Address)
+		}
+	}
+
+	return nil
+}
+
+// Allow unit tests to supply their own function for fetching HNSEndpoints.
+var hnsListEndpointRequest = hcsshim.HNSListEndpointRequest
+
+func findHNSEp(ip4, ip6 *net.IPNet) *hcsshim.HNSEndpoint {
+	hnsEndpoints, err := hnsListEndpointRequest()
```

Comment:
The call to `HNSListEndpointRequest` has no arguments and is unconditionally called, so could trivially be substituted with the `[]HNSEndpoint` value it would have returned. With just a little bit of refactoring, you could simply pass the mock `[]HNSEndpoint` slice directly to the code under test instead of having to mock out the function with a package-global variable.

```go
func findHNSEp(ip4, ip6 *net.IPNet, haystack []hcsshim.HNSEndpoint) *hcsshim.HNSEndpoint

func addEpToResolverImpl(netName, epName string, epIface *EndpointInterface, resolvers []*Resolver, hnsEPs []hcsshim.HNSEndpoint) error

func addEpToResolver(netName, epName string, epIface *EndpointInterface, resolvers []*Resolver) error {
	return addEpToResolverImpl(netName, epName, epIface, resolvers, hcsshim.HNSListEndpointRequest())
}
```

Reply:
Doh - that's how I did it initially, but I second-guessed myself into thinking you'd prefer it the other way! It's split again now.